Fix #12 and fix #14

alebg opened the pull request:
fix #12 and fix #14

alebg committed:
arangoDB Repository: major clean up of code, types, and functionality

- .use() method was removed as it is unnecessary in the new version of .useOrCreateDatabase()
- Two test methods were added, one for exporting back the env variables, the other to repristine the whole database
- Modified the core output port to match these changes; a type parameter was added to the port to better handle the Database type all around
- Modified DTOs to reflect changes; in particular, the Connection DTO
  for ArangoDB now returns both: the _system database, and the named
database we created; the _system one is used in some cases to create and
drop other databases
- Improved testing for arangodb-repository
- Removed duplicated .env.test file in the test folder, now there's a single one in the root of the project; note that the host port at test/docker-compose.yml  must manually match the ./.env.test port
- Added a mini migration file to play, but it's not being used yet

alebg committed:
project metadata: updated name of the project and README

alebg committed:
arangodb collection & document repository: major cleanup of code and typing, fixed and improved tests

- Working functionality for all of the main things: creating and using databases, creating collections (for both documents and edges), creating documents
- For testing purposes, now collections and documents can be created in arbitrary new databases
- Fixed all tests, now all pass
- Manual checking of collections by parsing the list of collections is replaced by the ".exists()" method
- Fixed typo: "aran<d>odb-document-repository.ts"

alebg committed:
package.json: docker commands for tests with the correct compose file

- Still the npm run test doesn't work, but tests pass in VSCode GUI

alebg committed:
arangodb prototyping test: assert that the DB is actually running; if this assertion is not there, this test will just always pass

alebg commented:
The CI will fail, but tests in the VSCode GUI pass (provided you have a test docker container running: `npm run docker:test-up` )

maany commented on package.json:
```diff
@@ -1,11 +1,12 @@
-    "docker:down": "docker-compose rm -fsv && rm .env",
+    "docker:dev": "cp .env.development .env && docker-compose --env-file .env.development up -d || docker compose --env-file .env.development up -d",
+    "docker:down": "docker-compose rm -fsv || docker compose rm -fsv && rm .env",
+    "docker:test-up": "cp .env.test .env && docker-compose -f test/docker-compose.yml --env-file .env.test up -d || docker compose -f test/docker-compose.yml --env-file .env.test up -d",
```
The CI has been configured to use `npm run docker:test` script. It fails because you've renamed the `docker:test` script to `docker:test-up`. You will have to adapt the github workflows to use these renamed scripts for bringing the containers up and down.

maany commented on src/lib/core/port/secondary/arangogb-document-repository-output-port.ts:
naming error: arangogb :)

maany commented on README.md:
````diff
@@ -24,14 +24,17 @@ This project uses [`next/font`](https://nextjs.org/docs/basic-features/font-opti
+For Testing, you can use the following command to start a test ArangoDB container (with a different port: `8530`):
 
-```
-npm run docker:test
````
Update the instructions for `npm run docker:test-up`and `npm run docker:test-down` here.

maany commented on src/lib/infrastructure/repository/arangodb-migrations/17-12-23-initialization.ts:
```diff
@@ -0,0 +1,36 @@
+    await db.collection('MetropolisEntity').drop()
+    return true
+  } catch (e) {
+    console.error(e)
```
should hard fail and not try to execute anything if an error occurs.

maany replied:
you can explicitly re-raise an error here and do something like a sys.exit(1) in python

maany commented on src/lib/infrastructure/repository/arangodb-migrations/17-12-23-initialization.ts:
```diff
@@ -0,0 +1,36 @@
+    await db.collection(`${db_name}Entity`).create()
+    return true
+  } catch (error) {
+    console.error(error)
```
hard error. See comment below

maany commented on src/lib/infrastructure/repository/arangodb-migrate.ts:
```diff
@@ -0,0 +1,12 @@
+import * as path from 'path'
+
+export async function migrate(db: Database) {
+  const migrationFiles = fs.readdirSync(path.join(__dirname, 'arangodb-migrations')).sort()
```
if sorting by name, then the migration file naming should not start with a date in dd/mm/yyyy format. You could use a yyyy/mm/dd date format instead.

maany commented on test/repository/arangodb/arangodb-repository.test.ts:
```diff
@@ -1,19 +1,107 @@
+        const system_db = connectionDTO.arangoSystemDB
+
+        // TODO: once we have a fixture for env variables, we can use the ones from there instead of this
+        const envVariablesDTO: ArangoDBEnvVariablesDTO = await arangoDBRepository._exportEnvironmentVariables()
```
the env gateway should be able to give you config from .env.test ?

maany reviewed (changes requested):
Looks good! Lots of clean up ;) I've left some comments

alebg committed:
tests and environment getway: code cleanup, fixed everything, fix #12, fix #14

- Fixed tests. Autogenerated database names were colliding when jest was running the tests in parallel. Now all database names are autogenerated unique strings

- Cleaned repository tests a bit more as well

- Removed redundant code and checks. E.g., now only arangoDBConfig() in the env gateway checks if env variables are missing, and throws an error msg saying precisely which ones. The compiler now just knows all env vars will be complete if no error was thrown

- Minor cleanup of older code

- Migrations are TODO, not important for now

alebg commented on test/api/delete-this.test.ts:
```diff
@@ -1,5 +1,3 @@
-    expect(true).toBeTruthy()
-  })
+test('dummy test', () => {
+  expect(true).toBe(true)
```
@maany sorry, deleted this one many commits ago but then noticed part of the CI tests fail without this. I put back a quick (but different) dummy test

maany commented on .github/workflows/repository-and-gateway.yml:
```diff
@@ -24,7 +24,7 @@ jobs:
         run: |
           docker pull arangodb:latest
-          npm run docker:test
+          npm run docker:test-up
```
is docker compose down the right way to bring down the containers or should it be npm run docker:test-down ?

alebg replied:
I think it should be with npm since our command also deletes volumes. Didn't notice

maany commented on .vscode/settings.json:
```diff
@@ -1,8 +1,5 @@
   "editor.defaultFormatter": "esbenp.prettier-vscode",
   "editor.formatOnSave": true,
-  "jest.jestCommandLine": "npm run test:repository -- --watch --env=jsdom",
-  "jest.autoRun": {
```
leave the onSave option ;) I use it

alebg replied:
How do I override it for me? I find it annoying

alebg replied:
nvm my local settings override it

maany commented on .env.development:
```diff
@@ -3,5 +3,5 @@ ARANGO_USER=root
 ARANGO_ROOT_PASSWORD=thereisnospoon
 ARANGO_URL=http://localhost
-ARANGO_PORT=8529
+ARANGO_PORT=8530
```
dev and test port are the same. Was this intentional?

alebg replied:
No, forgot to ask. What's a good practice? I didn't want them at the same port as prod. Is 3 different ports alright? Something like, prod at 8529 (arangodb default), dev at 8530 and test at 8531?

maany reviewed (commented):
Left a few comments

alebg committed:
minor changes: differentiated ports (prod, dev, test); CI kills containers with npm run command; put back vscode config
